src/main/java/teammates/storage/sqlsearch/SearchManager.java

```diff
@@ -0,0 +1,235 @@
+        } catch (SolrServerException e) {
+            log.severe(String.format(ERROR_DELETE_DOCUMENT, keys, e.getRootCause()), e);
+        } catch (IOException e) {
+            log.severe(String.format(ERROR_DELETE_DOCUMENT, keys, e.getCause()), e);
+        }
+    }
+
+    /**
+     * Resets the data for all collections if, and only if called during component
+     * tests.
+     */
+    public void resetCollections() {
+        if (client == null || !isResetAllowed) {
+            return;
+        }
+
+        try {
+            client.deleteByQuery(getCollectionName(), "*:*");
+            client.commit(getCollectionName());
+        } catch (SolrServerException e) {
+            log.severe(String.format(ERROR_RESET_COLLECTION, e.getRootCause()), e);
+        } catch (IOException e) {
+            log.severe(String.format(ERROR_RESET_COLLECTION, e.getCause()), e);
+        }
+    }
+
+    private String cleanSpecialChars(String queryString) {
+        String htmlTagStripPattern = "<[^>]*>";
+
+        // Solr special characters: + - && || ! ( ) { } [ ] ^ " ~ * ? : \ /
+        String res = queryString.replaceAll(htmlTagStripPattern, "")
+                .replace("\\", "\\\\")
+                .replace("+", "\\+")
+                .replace("-", "\\-")
+                .replace("&&", "\\&&")
+                .replace("||", "\\||")
+                .replace("!", "\\!")
+                .replace("(", "\\(")
+                .replace(")", "\\)")
+                .replace("{", "\\{")
+                .replace("}", "\\}")
+                .replace("[", "\\[")
+                .replace("]", "\\]")
+                .replace("^", "\\^")
+                .replace("~", "\\~")
+                .replace("?", "\\?")
+                .replace(":", "\\:")
+                .replace("/", "\\/");
+
+        // imbalanced double quotes are invalid
+        int count = StringUtils.countMatches(res, "\"");
+        if (count % 2 == 1) {
+            res = res.replace("\"", "");
+        }
+
+        // use exact match only when there's email-like input
+        if (res.contains("@") && count == 0) {
+            return "\"" + res + "\"";
+        } else {
+            return res;
+        }
+    }
+
+    abstract T getEntityFromDocument(SolrDocument document);
+
+    abstract void sortResult(List<T> result);
+
+    List<T> convertDocumentToEntities(List<SolrDocument> documents) {
+        if (documents == null) {
+            return new ArrayList<>();
+        }
+
+        List<T> result = new ArrayList<>();
+
+        for (SolrDocument document : documents) {
+            T entity = getEntityFromDocument(document);
+            if (entity == null) {
+                // search engine out of sync as SearchManager may fail to delete documents
+                // the chance is low and it is generally not a big problem
+                String id = (String) document.getFirstValue("id");
+                deleteDocuments(Collections.singletonList(id));
```

Comment:
Disable this for now in both the new and old implementations. Entity will be null here if document was for datastore (and vice versa). Do leave a comment explaining the rationale as well.